Bump body-parser to remove Node 10 deprecations

felixbuenemann opened the pull request:
This gets rid of `new Buffer()` deprecation warnings on Node v10 by bumping body-parser to the 1.18.3 version that was recently released by @dougwilson.

Also bumped qs to 6.5.2 to avoid duplicate versions.

See expressjs/body-parser#312 for details.

dougwilson commented:
Thanks! Can you please update the HISTORY.md file as well in the commit? Express 4.16 does not support Node.js 10 so probably will hold this for inclusion into the 4.17 release to get all the 10 issues resolved together :+1:

dougwilson reviewed (dismissed):
Update HISTORY file

dougwilson commented:
Here is an example dep update commit to help see what one looks like: https://github.com/expressjs/express/commit/cbaa04629a90aff76774b59cd055ff8c6e642030

dougwilson commented:
Oh yea, this also contains the http-errors bumps that I know there was a request to get into the current minor, so maybe we should still land as a patch. It looks like this also contains a module safer-buffer whereas the Express module set standardized on the module safe-buffer so that is kinda lame now we're going to pull in two different buffer shims into our install tree, lol

felixbuenemann commented:
The way the `History.md` file is structured it looks like it should be updated on release.

Adding to the 4.16.3 section would be wrong, since that version is already released and I cannot guess the new version and release date.

I can add a new section called `master` without a date that lists the dep changes on master since the last release.

dougwilson commented:
Sorry about that. Normally deps are never updated through PRs / non project members so it's not well documented. https://github.com/expressjs/express/commit/f448a9668515077937437d86e9e44421bcf2a093 is an example for the first note since a release.

felixbuenemann commented:
OK, I tried to reverse engineer the logic for deps and added entries based on the `HISTORY.md` from body-parser 1.18.3 and removed the deps entries that were already present in previous versions.

I'm not sure the `History.md` is complete though, for example there is no version that lists the bump to body-parser 1.18.2 which was at some point moved from dev dependency to dependency.

I also updated the commit message to `deps: body-parser@1.18.3` so it fits in with existing commits.

felixbuenemann commented:
> Oh yea, this also contains the http-errors bumps that I know there was a request to get into the current minor, so maybe we should still land as a patch. It looks like this also contains a module safer-buffer whereas the Express module set standardized on the module safe-buffer so that is kinda lame now we're going to pull in two different buffer shims into our install tree, lol

It looks like the safer-buffer dep is pulled in through `iconv-lite@0.4.23`, so unless they switch to a different package there's not much we can do about it.

Looks like you missed the `safer-buffer@'>= 2.1.2 < 3'` in the `HISTORY.md` of raw-body and body-parser. I'll update the `History.md` here to reflect that new dependency.

dougwilson commented:
P.S. maybe we should get this steps all written down to make it easier to make contributions against. Thanks for sticking with it. Let me know also if you would like someone else to make the commit too if you need. Also let me know if there is anything I can help with.

dougwilson commented:
> I'm not sure the History.md is complete though, for example there is no version that lists the bump to body-parser 1.18.2 which was at some point moved from dev dependency to dependency.

It is listed as the commit that make the change https://github.com/expressjs/express/commit/c0136d8b48dd3526c58b2ad8666fb4b12b55116c

> Add express.json and express.urlencoded to parse bodies

Before that body-parser was not included in Express so there were no commits upgrading the versions if that makes sense. The current format is just the commit message as the top level bullet point :+1:

felixbuenemann commented:
I was completely missing the point that the History.md top level bullet points where listing commit messages, while the second level where dependency changes.

I'll push an update to get rid of the extra top level bullets.

felixbuenemann committed:
deps: body-parser@1.18.3

closes #3655
closes #3706

felixbuenemann commented:
OK, I updated the History.md and also bumped qs to 6.5.2 in another commit so we don't depend on two different qs versions.

felixbuenemann commented:
Btw. only 3 packages on npm are currently depending on `safer-buffer`, but their README section  [Why not safe-buffer?](https://github.com/ChALkeR/safer-buffer#why-not-safe-buffer) lists some sane reasons why to prefer it over `safe-buffer`.

TL;DR `safer-buffer` prevents usage of the deprecated apis when required, while `safe-buffer` only polyfills the new features, but also shadows deprecation warnings.

dougwilson commented:
> It looks like the safer-buffer dep is pulled in through iconv-lite@0.4.23, so unless they switch to a different package there's not much we can do about it.

Right. That's why I'm actually thinking we should actually just lift all our other dependencies over to safer-buffer instead. Unless that package actually wants to change to safe-buffer haha. We initially went with safe-buffer when safer-buffer didn't exist. My understanding is that safer-buffer is actually better so it would make sense for us to move everything there anyways. Plus that would help a lot in regards to stability. Since that module depends on an extremely wide version range and our policy is to explicitly pin version on third party deps for stability and security, if we moved our tree to safer-buffer that would mitigate the issue since npm would end up using our pinned version as long as it was within that range.

felixbuenemann commented:
If you switch over basic-auth and express to safer-buffer you could get rid of an extra copy of `safe-bufer@5.1.1`, since other packages that need it resolve to 5.1.2. Otherwise it would make sense to bump to `safe-buffer@5.1.2` in both basic-auth and express before the next release.

Btw. what's the rationale for using such tight version control? – If dependencies are using semver it shouldn't be required and it's likely to cause unnecessary copies of packages due to minor version differences as well as making upgrades harder.

dougwilson commented:
> Btw. what's the rationale for using such tight version control? – If dependencies are using semver it shouldn't be required and it's likely to cause unnecessary copies of packages due to minor version differences as well as making upgrades harder.

There are several reason, all of which have burned Express in the past:

1. Semver is subjective to the author of the module on what they consider their API and what is a bug fix (all bug fixes are breaking changes). Even things like if hey consider breaking Node.js versions breaking or not would cause Express to stop working without notice.
2. User credentials get compromised and a new version can be published and suddenly pulled down by everyone installing Express. Npm will eventually pull that module from the registry, but in the meantime it did damage already. This is of no fault to the author and typically because they are not using a strong password, reusing passwords between sites and / or not using 2fa, none of which we can know from our side.
3. Module ownership changes hands. Even if we trusted who owns the module now, they can easily transfer ownership and now they have complete control over what Express users will start installing without any notice to us or any control. This ends up even going back to point 1 where the new owner has a different idea of what is breaking than the previous owner.

felixbuenemann commented:
Thanks for the explanation, those are all valid points!

Is there anything left to do here for me or is this good-to-merge?

dougwilson commented:
Yea, they are good, thanks for putting this together! I'm going to get the 4.17 branch rolling when I get to my computer tonight and get this and other changes merged in to start rolling on the 4.17 release, in which we can get Node 10 supported (I was waiting for Node 10 to actually be available on AppVeyor, but that hasn't happened yet and not sure what we're going to do about that yet), get moved over to safer-buffer, etc :+1:

dougwilson commented:
Nevermind Node.js 10 is available in AppVeyor (https://github.com/appveyor/ci/issues/2272) it just was not noted in their update log https://www.appveyor.com/updates/ which is what I monitor.

silverwind commented:
Looking forward to this being merged and released. Currently running any express application with `node --pending-deprecation` results in deprecations being logged because of outdated `body-parser`.
